File: shared/src/main/java/com/runtracker/shared/data/model/BodyAnalysis.kt
package com.runtracker.shared.data.model

import androidx.room.Entity
import androidx.room.PrimaryKey
import androidx.room.TypeConverters
import com.runtracker.shared.data.db.Converters

/**
 * Represents a body analysis scan with photo and AI-generated recommendations
 */
@Entity(tableName = "body_scans")
@TypeConverters(Converters::class)
data class BodyScan(
    @PrimaryKey(autoGenerate = true)
    val id: Long = 0,
    val timestamp: Long = System.currentTimeMillis(),
    val photoPath: String, // Local path to the saved photo
    val userGoal: FitnessGoal,
    val bodyType: BodyType,
    val estimatedBodyFatPercentage: Float? = null,
    val focusZones: List<BodyZone>,
    val overallScore: Int, // 0-100 fitness readiness score
    val muscleBalance: MuscleBalanceAssessment,
    val postureAssessment: PostureAssessment,
    val notes: String? = null
)

/**
 * User's primary fitness goal
 */
enum class FitnessGoal(val displayName: String, val description: String) {
    LOSE_WEIGHT("Lose Weight", "Focus on fat loss and cardio"),
    BUILD_MUSCLE("Build Muscle", "Focus on strength training and muscle growth"),
    TONE_UP("Tone & Define", "Reduce body fat while maintaining muscle"),
    IMPROVE_ENDURANCE("Improve Endurance", "Build cardiovascular fitness"),
    INCREASE_STRENGTH("Increase Strength", "Maximize power and lifting capacity"),
    GENERAL_FITNESS("General Fitness", "Overall health and wellness"),
    ATHLETIC_PERFORMANCE("Athletic Performance", "Sport-specific training"),
    BODY_RECOMPOSITION("Body Recomposition", "Lose fat and gain muscle simultaneously")
}

/**
 * Body type classification
 */
enum class BodyType(val displayName: String, val characteristics: String) {
    ECTOMORPH("Ectomorph", "Lean build, fast metabolism, difficulty gaining weight"),
    MESOMORPH("Mesomorph", "Athletic build, gains muscle easily, moderate metabolism"),
    ENDOMORPH("Endomorph", "Wider build, slower metabolism, gains weight easily"),
    ECTO_MESOMORPH("Ecto-Mesomorph", "Lean but can build muscle with effort"),
    ENDO_MESOMORPH("Endo-Mesomorph", "Athletic but prone to weight gain")
}

/**
 * Body zones that can be targeted for improvement
 */
enum class BodyZone(val displayName: String, val muscleGroups: List<String>) {
    UPPER_CHEST("Upper Chest", listOf("Pectoralis Major (Clavicular)")),
    LOWER_CHEST("Lower Chest", listOf("Pectoralis Major (Sternal)")),
    SHOULDERS("Shoulders", listOf("Anterior Deltoid", "Lateral Deltoid", "Posterior Deltoid")),
    UPPER_BACK("Upper Back", listOf("Trapezius", "Rhomboids", "Rear Deltoids")),
    LATS("Lats", listOf("Latissimus Dorsi")),
    LOWER_BACK("Lower Back", listOf("Erector Spinae")),
    BICEPS("Biceps", listOf("Biceps Brachii", "Brachialis")),
    TRICEPS("Triceps", listOf("Triceps Brachii")),
    FOREARMS("Forearms", listOf("Forearm Flexors", "Forearm Extensors")),
    ABS("Abs", listOf("Rectus Abdominis", "Transverse Abdominis")),
    OBLIQUES("Obliques", listOf("Internal Obliques", "External Obliques")),
    QUADS("Quadriceps", listOf("Rectus Femoris", "Vastus Lateralis", "Vastus Medialis")),
    HAMSTRINGS("Hamstrings", listOf("Biceps Femoris", "Semitendinosus", "Semimembranosus")),
    GLUTES("Glutes", listOf("Gluteus Maximus", "Gluteus Medius")),
    CALVES("Calves", listOf("Gastrocnemius", "Soleus")),
    CORE("Core", listOf("Deep Core Stabilizers", "Hip Flexors"))
}

/**
 * Priority level for focusing on a body zone
 */
enum class ZonePriority {
    HIGH, MEDIUM, LOW
}

/**
 * Recommendation for a specific body zone
 */
data class ZoneRecommendation(
    val zone: BodyZone,
    val priority: ZonePriority,
    val currentAssessment: String, // e.g., "Underdeveloped", "Well-balanced", "Strong"
    val recommendedExercises: List<String>,
    val weeklyFrequency: Int, // Recommended training days per week
    val notes: String? = null
)

/**
 * Muscle balance assessment
 */
data class MuscleBalanceAssessment(
    val overallBalance: BalanceLevel,
    val leftRightSymmetry: BalanceLevel,
    val frontBackBalance: BalanceLevel,
    val upperLowerBalance: BalanceLevel,
    val imbalances: List<MuscleImbalance>
)

enum class BalanceLevel(val displayName: String) {
    BALANCED("Balanced"),
    SLIGHT_IMBALANCE("Slight Imbalance"),
    MODERATE_IMBALANCE("Moderate Imbalance"),
    SIGNIFICANT_IMBALANCE("Significant Imbalance")
}

data class MuscleImbalance(
    val description: String,
    val affectedZones: List<BodyZone>,
    val correction: String
)

/**
 * Posture assessment
 */
data class PostureAssessment(
    val overallPosture: PostureLevel,
    val issues: List<PostureIssue>
)

enum class PostureLevel {
    EXCELLENT, GOOD, FAIR, POOR
}

data class PostureIssue(
    val type: PostureIssueType,
    val severity: IssueSeverity,
    val description: String,
    val exercises: List<String>
)

enum class PostureIssueType(val displayName: String) {
    FORWARD_HEAD("Forward Head Posture"),
    ROUNDED_SHOULDERS("Rounded Shoulders"),
    KYPHOSIS("Upper Back Rounding"),
    LORDOSIS("Excessive Lower Back Curve"),
    ANTERIOR_PELVIC_TILT("Anterior Pelvic Tilt"),
    POSTERIOR_PELVIC_TILT("Posterior Pelvic Tilt"),
    SCOLIOSIS("Lateral Spine Curvature"),
    UNEVEN_SHOULDERS("Uneven Shoulders"),
    UNEVEN_HIPS("Uneven Hips")
}

enum class IssueSeverity {
    MILD, MODERATE, SEVERE
}

/**
 * Complete analysis result with all recommendations
 */
data class BodyAnalysisResult(
    val scan: BodyScan,
    val zoneRecommendations: List<ZoneRecommendation>,
    val workoutRecommendations: List<WorkoutRecommendation>,
    val nutritionRecommendations: List<NutritionRecommendation>,
    val weeklyPlanSuggestion: WeeklyPlanSuggestion,
    val progressNotes: String? = null
)

/**
 * Workout recommendation based on analysis
 */
data class WorkoutRecommendation(
    val title: String,
    val description: String,
    val type: WorkoutRecommendationType,
    val frequency: String, // e.g., "3x per week"
    val duration: String, // e.g., "45-60 min"
    val exercises: List<RecommendedExercise>,
    val priority: Int // 1 = highest priority
)

enum class WorkoutRecommendationType {
    STRENGTH, HYPERTROPHY, CARDIO, HIIT, FLEXIBILITY, RECOVERY, POSTURE_CORRECTION
}

data class RecommendedExercise(
    val name: String,
    val sets: String,
    val reps: String,
    val notes: String? = null
)

/**
 * Nutrition recommendation
 */
data class NutritionRecommendation(
    val category: NutritionCategory,
    val title: String,
    val description: String,
    val tips: List<String>,
    val priority: Int
)

enum class NutritionCategory(val displayName: String, val icon: String) {
    PROTEIN("Protein", "🥩"),
    CARBS("Carbohydrates", "🍚"),
    FATS("Healthy Fats", "🥑"),
    HYDRATION("Hydration", "💧"),
    TIMING("Meal Timing", "⏰"),
    SUPPLEMENTS("Supplements", "💊"),
    CALORIES("Calorie Balance", "🔥"),
    MICRONUTRIENTS("Vitamins & Minerals", "🥗")
}

/**
 * Weekly training plan suggestion
 */
data class WeeklyPlanSuggestion(
    val totalTrainingDays: Int,
    val cardioSessions: Int,
    val strengthSessions: Int,
    val restDays: Int,
    val splitType: TrainingSplit,
    val dailyBreakdown: List<DayPlan>
)

enum class TrainingSplit(val displayName: String) {
    FULL_BODY("Full Body"),
    UPPER_LOWER("Upper/Lower"),
    PUSH_PULL_LEGS("Push/Pull/Legs"),
    BRO_SPLIT("Body Part Split"),
    CARDIO_STRENGTH("Cardio + Strength")
}

data class DayPlan(
    val dayOfWeek: String,
    val focus: String,
    val duration: String,
    val isRestDay: Boolean = false
)

/**
 * Comparison between two body scans
 */
data class BodyScanComparison(
    val currentScan: BodyScan,
    val previousScan: BodyScan?,
    val daysBetween: Int,
    val scoreChange: Int,
    val bodyFatChange: Float?,
    val improvedZones: List<BodyZone>,
    val needsMoreWorkZones: List<BodyZone>,
    val postureImprovement: Boolean,
    val overallProgress: ProgressAssessment,
    // Detailed improvement analysis
    val improvements: List<ImprovementItem> = emptyList(),
    val declines: List<ImprovementItem> = emptyList(),
    val unchanged: List<ImprovementItem> = emptyList(),
    val summaryMessage: String = "",
    val recommendations: List<String> = emptyList()
)

/**
 * Individual improvement/decline item for comparison
 */
data class ImprovementItem(
    val category: ImprovementCategory,
    val title: String,
    val description: String,
    val previousValue: String?,
    val currentValue: String,
    val changePercent: Float? = null,
    val isPositive: Boolean = true
)

enum class ImprovementCategory(val displayName: String, val icon: String) {
    OVERALL_SCORE("Overall Score", "📊"),
    BODY_FAT("Body Composition", "⚖️"),
    POSTURE("Posture", "🧍"),
    MUSCLE_BALANCE("Muscle Balance", "💪"),
    FOCUS_ZONES("Focus Areas", "🎯"),
    SYMMETRY("Symmetry", "↔️")
}

enum class ProgressAssessment(val displayName: String, val description: String) {
    EXCELLENT("Excellent Progress", "You're making great strides toward your goals!"),
    GOOD("Good Progress", "You're on the right track, keep it up!"),
    STEADY("Steady Progress", "Consistent effort is showing results."),
    SLOW("Slow Progress", "Consider adjusting your routine for better results."),
    PLATEAU("Plateau", "Time to shake things up with new challenges."),
    FIRST_SCAN("First Scan", "Great start! We'll track your progress from here.")
}

/**
 * Body scan with additional display information
 */
data class BodyScanWithDetails(
    val scan: BodyScan,
    val formattedDate: String,
    val goalDisplayName: String,
    val focusZoneCount: Int,
    val hasPostureIssues: Boolean
)

/**
 * Progress summary across all scans
 */
data class BodyProgressSummary(
    val totalScans: Int,
    val firstScanDate: Long?,
    val latestScanDate: Long?,
    val averageScore: Float,
    val bestScore: Int,
    val scoreImprovement: Int, // Latest vs first
    val consistentlyImprovedZones: List<BodyZone>,
    val persistentFocusZones: List<BodyZone>,
    val goalAchievementProgress: Float // 0-100%
)

File: iosApp/GoSteady/Sources/Models/BodyAnalysis.swift
import Foundation
import SwiftData

@Model
final class BodyScan {
    @Attribute(.unique) var id: UUID
    var timestamp: Date
    var photoPath: String
    var userGoal: FitnessGoal
    var bodyType: BodyType
    var estimatedBodyFatPercentage: Float?
    var focusZones: [BodyZone]
    var overallScore: Int
    var muscleBalance: MuscleBalanceAssessment
    var postureAssessment: PostureAssessment
    var notes: String?

    init(
        id: UUID = UUID(),
        timestamp: Date = Date(),
        photoPath: String,
        userGoal: FitnessGoal,
        bodyType: BodyType,
        estimatedBodyFatPercentage: Float? = nil,
        focusZones: [BodyZone],
        overallScore: Int,
        muscleBalance: MuscleBalanceAssessment,
        postureAssessment: PostureAssessment,
        notes: String? = nil
    ) {
        self.id = id
        self.timestamp = timestamp
        self.photoPath = photoPath
        self.userGoal = userGoal
        self.bodyType = bodyType
        self.estimatedBodyFatPercentage = estimatedBodyFatPercentage
        self.focusZones = focusZones
        self.overallScore = overallScore
        self.muscleBalance = muscleBalance
        self.postureAssessment = postureAssessment
        self.notes = notes
    }
}

enum FitnessGoal: String, Codable, CaseIterable, Identifiable {
    var id: String { rawValue }
    var icon: String {
        switch self {
        case .loseWeight: return "flame.fill"
        case .buildMuscle: return "dumbbell.fill"
        case .toneUp: return "figure.strengthtraining.traditional"
        case .improveEndurance: return "heart.fill"
        case .increaseStrength: return "bolt.fill"
        case .generalFitness: return "figure.run"
        case .athleticPerformance: return "sportscourt.fill"
        case .bodyRecomposition: return "figure.flexibility"
        }
    }
    //
    case loseWeight = "LOSE_WEIGHT"
    case buildMuscle = "BUILD_MUSCLE"
    case toneUp = "TONE_UP"
    case improveEndurance = "IMPROVE_ENDURANCE"
    case increaseStrength = "INCREASE_STRENGTH"
    case generalFitness = "GENERAL_FITNESS"
    case athleticPerformance = "ATHLETIC_PERFORMANCE"
    case bodyRecomposition = "BODY_RECOMPOSITION"

    var displayName: String {
        switch self {
        case .loseWeight: return "Lose Weight"
        case .buildMuscle: return "Build Muscle"
        case .toneUp: return "Tone & Define"
        case .improveEndurance: return "Improve Endurance"
        case .increaseStrength: return "Increase Strength"
        case .generalFitness: return "General Fitness"
        case .athleticPerformance: return "Athletic Performance"
        case .bodyRecomposition: return "Body Recomposition"
        }
    }

    var goalDescription: String {
        switch self {
        case .loseWeight: return "Focus on fat loss and cardio"
        case .buildMuscle: return "Focus on strength training and muscle growth"
        case .toneUp: return "Reduce body fat while maintaining muscle"
        case .improveEndurance: return "Build cardiovascular fitness"
        case .increaseStrength: return "Maximize power and lifting capacity"
        case .generalFitness: return "Overall health and wellness"
        case .athleticPerformance: return "Sport-specific training"
        case .bodyRecomposition: return "Lose fat and gain muscle simultaneously"
        }
    }
}

enum BodyType: String, Codable, CaseIterable {
    case ectomorph = "ECTOMORPH"
    case mesomorph = "MESOMORPH"
    case endomorph = "ENDOMORPH"
    case ectoMesomorph = "ECTO_MESOMORPH"
    case endoMesomorph = "ENDO_MESOMORPH"

    var displayName: String {
        switch self {
        case .ectomorph: return "Ectomorph"
        case .mesomorph: return "Mesomorph"
        case .endomorph: return "Endomorph"
        case .ectoMesomorph: return "Ecto-Mesomorph"
        case .endoMesomorph: return "Endo-Mesomorph"
        }
    }

    var characteristics: String {
        switch self {
        case .ectomorph: return "Lean build, fast metabolism, difficulty gaining weight"
        case .mesomorph: return "Athletic build, gains muscle easily, moderate metabolism"
        case .endomorph: return "Wider build, slower metabolism, gains weight easily"
        case .ectoMesomorph: return "Lean but can build muscle with effort"
        case .endoMesomorph: return "Athletic but prone to weight gain"
        }
    }
}

enum BodyZone: String, Codable, CaseIterable, Identifiable {
    var id: String { rawValue }
    //
    case upperChest = "UPPER_CHEST"
    case lowerChest = "LOWER_CHEST"
    case shoulders = "SHOULDERS"
    case upperBack = "UPPER_BACK"
    case lats = "LATS"
    case lowerBack = "LOWER_BACK"
    case biceps = "BICEPS"
    case triceps = "TRICEPS"
    case forearms = "FOREARMS"
    case abs = "ABS"
    case obliques = "OBLIQUES"
    case quads = "QUADS"
    case hamstrings = "HAMSTRINGS"
    case glutes = "GLUTES"
    case calves = "CALVES"
    case core = "CORE"

    var displayName: String {
        switch self {
        case .upperChest: return "Upper Chest"
        case .lowerChest: return "Lower Chest"
        case .shoulders: return "Shoulders"
        case .upperBack: return "Upper Back"
        case .lats: return "Lats"
        case .lowerBack: return "Lower Back"
        case .biceps: return "Biceps"
        case .triceps: return "Triceps"
        case .forearms: return "Forearms"
        case .abs: return "Abs"
        case .obliques: return "Obliques"
        case .quads: return "Quadriceps"
        case .hamstrings: return "Hamstrings"
        case .glutes: return "Glutes"
        case .calves: return "Calves"
        case .core: return "Core"
        }
    }

    var muscleGroups: [String] {
        switch self {
        case .upperChest: return ["Pectoralis Major (Clavicular)"]
        case .lowerChest: return ["Pectoralis Major (Sternal)"]
        case .shoulders: return ["Anterior Deltoid", "Lateral Deltoid", "Posterior Deltoid"]
        case .upperBack: return ["Trapezius", "Rhomboids", "Rear Deltoids"]
        case .lats: return ["Latissimus Dorsi"]
        case .lowerBack: return ["Erector Spinae"]
        case .biceps: return ["Biceps Brachii", "Brachialis"]
        case .triceps: return ["Triceps Brachii"]
        case .forearms: return ["Forearm Flexors", "Forearm Extensors"]
        case .abs: return ["Rectus Abdominis", "Transverse Abdominis"]
        case .obliques: return ["Internal Obliques", "External Obliques"]
        case .quads: return ["Rectus Femoris", "Vastus Lateralis", "Vastus Medialis"]
        case .hamstrings: return ["Biceps Femoris", "Semitendinosus", "Semimembranosus"]
        case .glutes: return ["Gluteus Maximus", "Gluteus Medius"]
        case .calves: return ["Gastrocnemius", "Soleus"]
        case .core: return ["Deep Core Stabilizers", "Hip Flexors"]
        }
    }
}

enum ZonePriority: String, Codable, CaseIterable {
    case high = "HIGH"
    case medium = "MEDIUM"
    case low = "LOW"
}

struct ZoneRecommendation: Codable, Hashable {
    var zone: BodyZone
    var priority: ZonePriority
    var currentAssessment: String
    var recommendedExercises: [String]
    var weeklyFrequency: Int
    var notes: String?
}

struct MuscleBalanceAssessment: Codable, Hashable {
    var overallBalance: BalanceLevel
    var leftRightSymmetry: BalanceLevel
    var frontBackBalance: BalanceLevel
    var upperLowerBalance: BalanceLevel
    var imbalances: [MuscleImbalance]

    init(
        overallBalance: BalanceLevel = .balanced,
        leftRightSymmetry: BalanceLevel = .balanced,
        frontBackBalance: BalanceLevel = .balanced,
        upperLowerBalance: BalanceLevel = .balanced,
        imbalances: [MuscleImbalance] = []
    ) {
        self.overallBalance = overallBalance
        self.leftRightSymmetry = leftRightSymmetry
        self.frontBackBalance = frontBackBalance
        self.upperLowerBalance = upperLowerBalance
        self.imbalances = imbalances
    }
}

enum BalanceLevel: String, Codable, CaseIterable {
    case balanced = "BALANCED"
    case slightImbalance = "SLIGHT_IMBALANCE"
    case moderateImbalance = "MODERATE_IMBALANCE"
    case significantImbalance = "SIGNIFICANT_IMBALANCE"

    var displayName: String {
        switch self {
        case .balanced: return "Balanced"
        case .slightImbalance: return "Slight Imbalance"
        case .moderateImbalance: return "Moderate Imbalance"
        case .significantImbalance: return "Significant Imbalance"
        }
    }
}

struct MuscleImbalance: Codable, Hashable, Identifiable {
    var id: String { imbalanceDescription }
    var imbalanceDescription: String
    var affectedZones: [BodyZone]
    var correction: String
}

struct PostureAssessment: Codable, Hashable {
    var overallPosture: PostureLevel
    var issues: [PostureIssue]

    init(overallPosture: PostureLevel = .good, issues: [PostureIssue] = []) {
        self.overallPosture = overallPosture
        self.issues = issues
    }
}

enum PostureLevel: String, Codable, CaseIterable {
    case excellent = "EXCELLENT"
    case good = "GOOD"
    case fair = "FAIR"
    case poor = "POOR"

    var displayName: String {
        switch self {
        case .excellent: return "Excellent"
        case .good: return "Good"
        case .fair: return "Fair"
        case .poor: return "Poor"
        }
    }
}

struct PostureIssue: Codable, Hashable, Identifiable {
    var id: String { "\(type.rawValue)_\(severity.rawValue)" }
    var type: PostureIssueType
    var severity: IssueSeverity
    var issueDescription: String
    var exercises: [String]
}

enum PostureIssueType: String, Codable, CaseIterable {
    case forwardHead = "FORWARD_HEAD"
    case roundedShoulders = "ROUNDED_SHOULDERS"
    case kyphosis = "KYPHOSIS"
    case lordosis = "LORDOSIS"
    case anteriorPelvicTilt = "ANTERIOR_PELVIC_TILT"
    case posteriorPelvicTilt = "POSTERIOR_PELVIC_TILT"
    case scoliosis = "SCOLIOSIS"
    case unevenShoulders = "UNEVEN_SHOULDERS"
    case unevenHips = "UNEVEN_HIPS"

    var displayName: String {
        switch self {
        case .forwardHead: return "Forward Head Posture"
        case .roundedShoulders: return "Rounded Shoulders"
        case .kyphosis: return "Upper Back Rounding"
        case .lordosis: return "Excessive Lower Back Curve"
        case .anteriorPelvicTilt: return "Anterior Pelvic Tilt"
        case .posteriorPelvicTilt: return "Posterior Pelvic Tilt"
        case .scoliosis: return "Lateral Spine Curvature"
        case .unevenShoulders: return "Uneven Shoulders"
        case .unevenHips: return "Uneven Hips"
        }
    }
}

enum IssueSeverity: String, Codable, CaseIterable {
    case mild = "MILD"
    case moderate = "MODERATE"
    case severe = "SEVERE"

    var displayName: String {
        switch self {
        case .mild: return "Mild"
        case .moderate: return "Moderate"
        case .severe: return "Severe"
        }
    }
}

struct BodyAnalysisReport: Codable {
    var scan: BodyScanData
    var zoneRecommendations: [ZoneRecommendation]
    var workoutRecommendations: [WorkoutRecommendation]
    var nutritionRecommendations: [NutritionRecommendation]
    var weeklyPlanSuggestion: WeeklyPlanSuggestion
    var progressNotes: String?
}

struct BodyScanData: Codable {
    var id: UUID
    var timestamp: Date
    var photoPath: String
    var userGoal: FitnessGoal
    var bodyType: BodyType
    var estimatedBodyFatPercentage: Float?
    var focusZones: [BodyZone]
    var overallScore: Int
    var muscleBalance: MuscleBalanceAssessment
    var postureAssessment: PostureAssessment
    var notes: String?
}

struct WorkoutRecommendation: Codable, Hashable {
    var title: String
    var recommendationDescription: String
    var type: WorkoutRecommendationType
    var frequency: String
    var duration: String
    var exercises: [RecommendedExercise]
    var priority: Int
}

enum WorkoutRecommendationType: String, Codable, CaseIterable {
    case strength = "STRENGTH"
    case hypertrophy = "HYPERTROPHY"
    case cardio = "CARDIO"
    case hiit = "HIIT"
    case flexibility = "FLEXIBILITY"
    case recovery = "RECOVERY"
    case postureCorrection = "POSTURE_CORRECTION"
}

struct RecommendedExercise: Codable, Hashable {
    var name: String
    var sets: String
    var reps: String
    var notes: String?
}

struct NutritionRecommendation: Codable, Hashable {
    var category: NutritionCategory
    var title: String
    var recommendationDescription: String
    var tips: [String]
    var priority: Int
}

enum NutritionCategory: String, Codable, CaseIterable {
    case protein = "PROTEIN"
    case carbs = "CARBS"
    case fats = "FATS"
    case hydration = "HYDRATION"
    case timing = "TIMING"
    case supplements = "SUPPLEMENTS"
    case calories = "CALORIES"
    case micronutrients = "MICRONUTRIENTS"

    var displayName: String {
        switch self {
        case .protein: return "Protein"
        case .carbs: return "Carbohydrates"
        case .fats: return "Healthy Fats"
        case .hydration: return "Hydration"
        case .timing: return "Meal Timing"
        case .supplements: return "Supplements"
        case .calories: return "Calorie Balance"
        case .micronutrients: return "Vitamins & Minerals"
        }
    }
}

struct WeeklyPlanSuggestion: Codable, Hashable {
    var totalTrainingDays: Int
    var cardioSessions: Int
    var strengthSessions: Int
    var restDays: Int
    var splitType: TrainingSplit
    var dailyBreakdown: [DayPlan]
}

enum TrainingSplit: String, Codable, CaseIterable {
    case fullBody = "FULL_BODY"
    case upperLower = "UPPER_LOWER"
    case pushPullLegs = "PUSH_PULL_LEGS"
    case broSplit = "BRO_SPLIT"
    case cardioStrength = "CARDIO_STRENGTH"

    var displayName: String {
        switch self {
        case .fullBody: return "Full Body"
        case .upperLower: return "Upper/Lower"
        case .pushPullLegs: return "Push/Pull/Legs"
        case .broSplit: return "Body Part Split"
        case .cardioStrength: return "Cardio + Strength"
        }
    }
}

struct DayPlan: Codable, Hashable {
    var dayOfWeek: String
    var focus: String
    var duration: String
    var isRestDay: Bool

    init(dayOfWeek: String, focus: String, duration: String, isRestDay: Bool = false) {
        self.dayOfWeek = dayOfWeek
        self.focus = focus
        self.duration = duration
        self.isRestDay = isRestDay
    }
}

struct BodyScanComparison: Codable {
    var currentScan: BodyScanData
    var previousScan: BodyScanData?
    var daysBetween: Int
    var scoreChange: Int
    var bodyFatChange: Float?
    var improvedZones: [BodyZone]
    var needsMoreWorkZones: [BodyZone]
    var postureImprovement: Bool
    var overallProgress: ProgressAssessment
    var improvements: [ImprovementItem]
    var declines: [ImprovementItem]
    var unchanged: [ImprovementItem]
    var summaryMessage: String
    var recommendations: [String]
}

struct ImprovementItem: Codable, Hashable {
    var category: ImprovementCategory
    var title: String
    var itemDescription: String
    var previousValue: String?
    var currentValue: String
    var changePercent: Float?
    var isPositive: Bool

    init(
        category: ImprovementCategory,
        title: String,
        itemDescription: String,
        previousValue: String? = nil,
        currentValue: String,
        changePercent: Float? = nil,
        isPositive: Bool = true
    ) {
        self.category = category
        self.title = title
        self.itemDescription = itemDescription
        self.previousValue = previousValue
        self.currentValue = currentValue
        self.changePercent = changePercent
        self.isPositive = isPositive
    }
}

enum ImprovementCategory: String, Codable, CaseIterable {
    case overallScore = "OVERALL_SCORE"
    case bodyFat = "BODY_FAT"
    case posture = "POSTURE"
    case muscleBalance = "MUSCLE_BALANCE"
    case focusZones = "FOCUS_ZONES"
    case symmetry = "SYMMETRY"

    var displayName: String {
        switch self {
        case .overallScore: return "Overall Score"
        case .bodyFat: return "Body Composition"
        case .posture: return "Posture"
        case .muscleBalance: return "Muscle Balance"
        case .focusZones: return "Focus Areas"
        case .symmetry: return "Symmetry"
        }
    }
}

enum ProgressAssessment: String, Codable, CaseIterable {
    case excellent = "EXCELLENT"
    case good = "GOOD"
    case steady = "STEADY"
    case slow = "SLOW"
    case plateau = "PLATEAU"
    case firstScan = "FIRST_SCAN"

    var displayName: String {
        switch self {
        case .excellent: return "Excellent Progress"
        case .good: return "Good Progress"
        case .steady: return "Steady Progress"
        case .slow: return "Slow Progress"
        case .plateau: return "Plateau"
        case .firstScan: return "First Scan"
        }
    }

    var assessmentDescription: String {
        switch self {
        case .excellent: return "You're making great strides toward your goals!"
        case .good: return "You're on the right track, keep it up!"
        case .steady: return "Consistent effort is showing results."
        case .slow: return "Consider adjusting your routine for better results."
        case .plateau: return "Time to shake things up with new challenges."
        case .firstScan: return "Great start! We'll track your progress from here."
        }
    }
}

struct BodyScanWithDetails: Codable {
    var scan: BodyScanData
    var formattedDate: String
    var goalDisplayName: String
    var focusZoneCount: Int
    var hasPostureIssues: Bool
}

struct BodyProgressSummary: Codable {
    var totalScans: Int
    var firstScanDate: Date?
    var latestScanDate: Date?
    var averageScore: Float
    var bestScore: Int
    var scoreImprovement: Int
    var consistentlyImprovedZones: [BodyZone]
    var persistentFocusZones: [BodyZone]
    var goalAchievementProgress: Float
}
